File: src/main/java/thkoeln/dungeon/planet/domain/Planet.java
package thkoeln.dungeon.planet.domain;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.text.WordUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.persistence.*;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Entity
@Getter
public class Planet {
    @Id
    private final UUID id = UUID.randomUUID();

    @Setter
    private String name;

    @Setter
    @Getter ( AccessLevel.NONE )
    private Boolean spacestation = Boolean.FALSE;
    public Boolean isSpaceStation() { return spacestation; }

    @OneToOne ( cascade = CascadeType.MERGE)
    @Setter ( AccessLevel.PROTECTED )
    private Planet northNeighbour;
    @OneToOne ( cascade = CascadeType.MERGE)
    @Setter ( AccessLevel.PROTECTED )
    private Planet eastNeighbour;
    @OneToOne ( cascade = CascadeType.MERGE)
    @Setter ( AccessLevel.PROTECTED )
    private Planet southNeighbour;
    @OneToOne ( cascade = CascadeType.MERGE)
    @Setter ( AccessLevel.PROTECTED )
    private Planet westNeighbour;

    @Transient
    private Logger logger = LoggerFactory.getLogger( Planet.class );

    /**
     * A neighbour relationship is always set on BOTH sides.
     * @param otherPlanet
     * @param direction
     */
    public void defineNeighbour(Planet otherPlanet, CompassDirection direction ) {
        if ( otherPlanet == null ) throw new PlanetException( "Cannot establish neighbouring relationship with null planet!" ) ;
        try {
            Method otherGetter = neighbouringGetter( direction.getOppositeDirection() );
            Method setter = neighbouringSetter( direction );
            setter.invoke(this, otherPlanet );
            Planet remoteNeighbour = (Planet) otherGetter.invoke( otherPlanet );
            if ( !this.equals( remoteNeighbour ) ) {
                Method otherSetter = neighbouringSetter( direction.getOppositeDirection() );
                otherSetter.invoke( otherPlanet, this );
            }
        }
        catch ( IllegalAccessException | InvocationTargetException | NoSuchMethodException e ) {
            throw new PlanetException( "Something went wrong that should not have happened ..." + e.getStackTrace() );
        }
        logger.info( "Established neighbouring relationship between planet '" + this + "' and '" + otherPlanet + "'." );
    }


    protected Method neighbouringGetter( CompassDirection direction ) throws NoSuchMethodException {
        String name = "get" + WordUtils.capitalize( String.valueOf( direction ) ) + "Neighbour";
        return this.getClass().getDeclaredMethod( name );
    }


    protected Method neighbouringSetter( CompassDirection direction ) throws NoSuchMethodException {
        String name = "set" + WordUtils.capitalize( String.valueOf( direction ) ) + "Neighbour";
        return this.getClass().getDeclaredMethod( name, new Class[]{ this.getClass() } );
    }


    public List<Planet> allNeighbours() {
        List<Planet> allNeighbours = new ArrayList<>();
        if ( getNorthNeighbour() != null ) allNeighbours.add( getNorthNeighbour() );
        if ( getWestNeighbour() != null ) allNeighbours.add( getWestNeighbour() );
        if ( getEastNeighbour() != null ) allNeighbours.add( getEastNeighbour() );
        if ( getSouthNeighbour() != null ) allNeighbours.add( getSouthNeighbour() );
        return allNeighbours;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Planet)) return false;
        Planet planet = (Planet) o;
        return Objects.equals(id, planet.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return getName() + " (" + getId() + ")";
    }
}

File: src/main/java/thkoeln/dungeon/planet/domain/Planet.kt
package thkoeln.dungeon.planet.domain

import lombok.*
import org.apache.commons.text.WordUtils
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*
import javax.persistence.*

@Entity
@Getter
class Planet {
    @Id
    val id = UUID.randomUUID()

    @Setter
    var name: String? = null

    @Setter
    @Getter(AccessLevel.NONE)
    val isSpaceStation = java.lang.Boolean.FALSE

    @OneToOne(cascade = [CascadeType.MERGE])
    @Setter(AccessLevel.PROTECTED)
    val northNeighbour: Planet? = null

    @OneToOne(cascade = [CascadeType.MERGE])
    @Setter(AccessLevel.PROTECTED)
    val eastNeighbour: Planet? = null

    @OneToOne(cascade = [CascadeType.MERGE])
    @Setter(AccessLevel.PROTECTED)
    val southNeighbour: Planet? = null

    @OneToOne(cascade = [CascadeType.MERGE])
    @Setter(AccessLevel.PROTECTED)
    val westNeighbour: Planet? = null

    @Transient
    private val logger = LoggerFactory.getLogger(Planet::class.java)

    /**
     * A neighbour relationship is always set on BOTH sides.
     * @param otherPlanet
     * @param direction
     */
    fun defineNeighbour(otherPlanet: Planet?, direction: CompassDirection) {
        if (otherPlanet == null) throw PlanetException("Cannot establish neighbouring relationship with null planet!")
        try {
            val otherGetter = neighbouringGetter(direction.oppositeDirection)
            val setter = neighbouringSetter(direction)
            setter.invoke(this, otherPlanet)
            val remoteNeighbour = otherGetter.invoke(otherPlanet) as Planet
            if (this != remoteNeighbour) {
                val otherSetter = neighbouringSetter(direction.oppositeDirection)
                otherSetter.invoke(otherPlanet, this)
            }
        } catch (e: IllegalAccessException) {
            throw PlanetException("Something went wrong that should not have happened ..." + e.stackTrace)
        } catch (e: InvocationTargetException) {
            throw PlanetException("Something went wrong that should not have happened ..." + e.stackTrace)
        } catch (e: NoSuchMethodException) {
            throw PlanetException("Something went wrong that should not have happened ..." + e.stackTrace)
        }
        logger.info("Established neighbouring relationship between planet '$this' and '$otherPlanet'.")
    }

    @Throws(NoSuchMethodException::class)
    fun neighbouringGetter(direction: CompassDirection?): Method {
        val name = "get" + WordUtils.capitalize(direction.toString()) + "Neighbour"
        return this.javaClass.getDeclaredMethod(name)
    }

    @Throws(NoSuchMethodException::class)
    fun neighbouringSetter(direction: CompassDirection?): Method {
        val name = "set" + WordUtils.capitalize(direction.toString()) + "Neighbour"
        return this.javaClass.getDeclaredMethod(name, *arrayOf<Class<*>>(this.javaClass))
    }

    fun allNeighbours(): List<Planet> {
        val allNeighbours: MutableList<Planet> = ArrayList()
        if (this.northNeighbour != null) allNeighbours.add(this.northNeighbour)
        if (this.westNeighbour != null) allNeighbours.add(this.westNeighbour)
        if (this.eastNeighbour != null) allNeighbours.add(this.eastNeighbour)
        if (this.southNeighbour != null) allNeighbours.add(this.southNeighbour)
        return allNeighbours
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Planet) return false
        return id == o.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    override fun toString(): String {
        return this.name + " (" + this.id + ")"
    }
}